src/drivers/uavcan_v1/Subscribers/DynamicPortSubscriber.hpp

```diff
@@ -60,30 +60,37 @@ class UavcanDynamicPortSubscriber : public UavcanBaseSubscriber
 
 	void updateParam()
 	{
-		char uavcan_param[90];
-		sprintf(uavcan_param, "uavcan.sub.%s.%d.id", _subject_name, _instance);
+		SubjectSubscription *curSubj = &_subj_sub;
 
-		// Set _port_id from _uavcan_param
-		uavcan_register_Value_1_0 value;
-		_param_manager.GetParamByName(uavcan_param, value);
-		int32_t new_id = value.integer32.value.elements[0];
+		while (curSubj != NULL) {
+			char uavcan_param[90];
+			snprintf(uavcan_param, sizeof(uavcan_param), "uavcan.sub.%s.%d.id", curSubj->_subject_name, _instance);
 
-		if (_port_id != new_id) {
-			if (new_id == CANARD_PORT_ID_UNSET) {
-				// Cancel subscription
-				unsubscribe();
+			// Set _port_id from _uavcan_param
+			uavcan_register_Value_1_0 value;
+			_param_manager.GetParamByName(uavcan_param, value);
+			int32_t new_id = value.integer32.value.elements[0];
```

Comment:
I [commented on this in one of the earlier pull requests](https://github.com/PX4/PX4-Autopilot/pull/16889#discussion_r582032230) but it somehow slipped through the cracks: these should be `natural16` (which is `uint16_t`), not `integer32`.

Reply:
Thanks fixed that one

Reply:
This is not the correct fix because now you are just referencing the same value as a different type. A correct fix would be to add a converter lambda to the `UavcanParamManager` as I suggested in the linked above thread.

Reply:
Okay I will revert my attempted fix then, this PR is not changing the param system as well I just moved the definition from the class to another.

To apply a correct fix like you've described above would require a rewrite of the param system but let's do that in a later iteration.